opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/Wise.kt

```diff
@@ -0,0 +1,563 @@
+        val data = epochDay + applicationName
+
+        val mac = Mac.getInstance("HmacSHA256")
+        val secret_key = SecretKeySpec(apiKey.toByteArray(charset("UTF-8")), "HmacSHA256")
+        mac.init(secret_key)
+        val signature= bytesToHex(mac.doFinal(data.toByteArray(charset("UTF-8"))))
+        return "$apiKeyId:$signature"
+    }
+    private val hexArray = "0123456789abcdef".toCharArray()
+
+    fun bytesToHex(bytes: ByteArray): String {
+        val hexChars = CharArray(bytes.size * 2)
+        for (j in bytes.indices) {
+            var v = (bytes[j] and 0xFF.toByte()).toInt()
+            if ( v < 0) v = 256+v
+
+            hexChars[j * 2] = hexArray[v ushr 4]
+            hexChars[j * 2 + 1] = hexArray[v and 0x0F]
+        }
+        return String(hexChars)
+    }
+}
+
+open class Wise : OkHttpBaseApi() {
+    protected lateinit var baseurl: String
+    protected lateinit var imgurl: String
+    protected var searchLocations: List<Pair<String,String>> = listOf(Pair("0","Own branch"))
+    protected var homeBranch: Long? = -1
+    protected lateinit var libraryData: JSONObject
+    protected lateinit var queryString: String
+    protected lateinit var scope: String
+    protected var mediaTypeQS: String = ""
+    protected var vcgrpt: String = ""
+    protected lateinit var holdParameters: String
+    protected var pickupBranch: Long? = -1
+
+    protected val ENCODING = "UTF-8"
+
+    private val mediatypes = mapOf(
+//            null to SearchResult.MediaType.UNKNOWN
+            "CD" to SearchResult.MediaType.CD,
+            "DVD" to SearchResult.MediaType.DVD,
+            "DVR" to SearchResult.MediaType.GAME_CONSOLE,
+            "CSP" to SearchResult.MediaType.GAME_CONSOLE,
+            "GBK" to SearchResult.MediaType.AUDIOBOOK,
+            "BOE" to SearchResult.MediaType.BOOK
+    )
+
+    private var itemstatuses: Map<String, String>? = null
+    private var holdstatuses: Map<String, String>? = null
+
+    override fun init(library: Library, factory: HttpClientFactory, debug: Boolean) {
+        http_interceptor = WiseKeyInterceptor()
+        super.init(library, factory, debug)
+        libraryData = library.data
+        homeBranch = library.libraryId
+        baseurl = libraryData.getString("baseurl")
+        imgurl = libraryData.getString("imgurl")
+        val cats = libraryData.optJSONArray("search-locations")
+        if ( cats != null)  {
+            val catsList =  List(cats.length(),cats::getJSONObject)
+            searchLocations = emptyList()
+            catsList.forEach {
+                json -> val key = json.names()[0];  searchLocations = searchLocations + Pair(key as String, json.get(key) as String)
+            }
+        }
+        if ( libraryData.has("api-key-id") ) {
+            apiKeyId = libraryData.getString("api-key-id")
+            apiKey = libraryData.getString("api-key")
+            applicationName = libraryData.getString("app-name")
+        }
```

Comment:
What happens in the `else` case? These are lazyinit properties, do things just crash?

Reply:
will check

Reply:
These `api` settings items are always needed. I will remove if-check